src/crawler_utils/scraping_utils.py

```diff
@@ -40,7 +40,7 @@ def open_url(driver, url):
     def bs4_scrape(soup, selector, keyword, attr=False):
```

Comment:
Can you try to define this function like this:
```python
import bs4

def bs4_scrape(soup, selector, keyword, attr=None):
    """
    Scrapes data from a BeautifulSoup object based on the provided selector, keyword, and optional attribute.

    This function simplifies web scraping using BeautifulSoup by providing a flexible interface to find and extract
    data from HTML or XML documents. It can search for elements using different selectors, such as 'findAll',
    'find_all', 'find', 'find_next', or 'findNext', and optionally filter by attributes.

    :param soup: A BeautifulSoup object or Tag to search within.
    :param selector: The selection method, e.g., 'findAll', 'find_all', 'find', 'find_next', or 'findNext'.
    :param keyword: The HTML or XML element name or tag to search for.
    :param attr: (Optional) A dictionary of attributes and values to filter elements by.
    :return: The scraped data as a BeautifulSoup ResultSet or Tag, or None if no data is found.
    :raises ValueError: If an invalid object is provided as 'soup'.
    """
    if not isinstance(soup, (bs4.BeautifulSoup, bs4.element.Tag, bs4.element.ResultSet)):
        raise ValueError(f"Invalid object type provided as 'soup': {type(soup)}. Expected: bs4.BeautifulSoup or Tag.")

    try:
        if selector in {'findAll', 'find_all', 'find', 'find_next', 'findNext'}:
            if attr:
                data = getattr(soup, selector)(keyword, attrs=attr)
            else:
                data = getattr(soup, selector)(keyword)
            return data

    except Exception as exc:
        print(f"Error during scraping: {exc}")

    return None
```

This will avoid repetitive code

Reply:
Changes incorporated.